cowprotocol/services
# Price competition bang for buck

# Description
We started to allow external solvers to provide quotes and incentivized them by giving them rewards for every trade where they reported the best quote.
At the moment we rank qoutes by the highest reported `out_amount` without taking gas costs of complex routes into account. Because solvers are now competing on the highest `out_amount` they started to propose more and more complex routes which would end up hurting the user because complex routes require a lot of gas to execute which would end up being deducted from the user in form of fees.
In summary the current situation leads to the UI showing very high fees which leads to fewer trades being placed.

# Changes
Instead of most `out_amount` we now rank by `bang-for-buck` (i.e. `out_amount - gas_fees`). To do that our price estimators now also need to be aware of the native price for the `out_token`. That way we can rank estimates by the total `eth_received`.

In order to keep the interface unchanged I decided to make the regular price estimator use a native price estimator under the hood.
Because the `CompetitionPriceEstimator` was recently refactored to be usable for regular and native price estimates the native price estimator now also has the field `native`. This is obviously not great but at least you don't have to pass an actual native price estimator and it also ends up not being used.

## How to test
Not sure how doable it is to set up an `e2e` test for this but I should at least add a unit test using mocks that verifies that we now pick best "bang-for-buck".
